pkg/controllermanager/controller/shoot/shoot_conditions_control.go

```diff
@@ -38,26 +38,30 @@ func (c *Controller) shootConditionsAdd(obj interface{}) {
 	c.shootConditionsQueue.Add(key)
 }
 
-func (c *Controller) filterSeedForShootConditions(obj, oldObj, _ client.Object, deleted bool) bool {
+func (c *Controller) filterSeedForShootConditions(obj, oldObj, _ client.Object, _ bool) bool {
 	seed, ok := obj.(*gardencorev1beta1.Seed)
 	if !ok {
 		return false
 	}
 
-	if oldObj != nil {
-		oldSeed, ok := oldObj.(*gardencorev1beta1.Seed)
-		if !ok {
-			return false
-		}
+	// We want to enqueue in case of deletion events to remove conditions.
+	// We want to enqueue in case of add events as they can indicate restarts or reflector relists.
+	if oldObj == nil {
+		return true
+	}
 
-		if !apiequality.Semantic.DeepEqual(seed.Status.Conditions, oldSeed.Status.Conditions) {
-			logger.Logger.Debugf("Seed %s conditions changed", seed.Name)
-			return true
-		}
+	oldSeed, ok := oldObj.(*gardencorev1beta1.Seed)
+	if !ok {
+		return false
+	}
+
+	if !apiequality.Semantic.DeepEqual(seed.Status.Conditions, oldSeed.Status.Conditions) {
+		logger.Logger.Debugf("Seed %s conditions changed", seed.Name)
+		return true
 	}
 
-	if deleted {
-		logger.Logger.Debugf("Seed %s deleted", seed.Name)
+	// We want to enqueue on periodic cache resync events to catch up if we missed updates.
+	if seed.ResourceVersion == oldSeed.ResourceVersion {
```

Comment:
Considering my findings in https://github.com/gardener/gardener/issues/5295#issuecomment-1018285002, I'm asking myself if it wouldn't be better to handling resync events in the seed event handler (seeds are kind of the "owned objects" here) and instead add an update handler to the shoot informer (shoots are the "controlled objects" here), which respects periodic resyncs.
The advantage would be that resync events also cause the controller to remove seed conditions from shoots where the seed has been deleted (relevant for cases where the seed informer lags behind).

WDYT?

Reply:
> and instead add an update handler to the shoot informer

Hm, let's suppose we missed an update or the seed informer lags behind, don't we want to have a resync on seeds rather than on shoots?
How would a resync on shoots let us consider the "right' conditions of the seed?

Reply:
I was thinking about the following case:
- seed informer lags behind
- seed is deleted
- watch connection eventually breaks and seed informer does a full relist

I thought, that a shoot resync helps to mitigate missing seed delete events.

Now I realize that neither a seed nor shoot resync help in this case:
A resync of the seed informer doesn't help, because it doesn't refresh the cache's state. Even if the predicate evaluates to `true` and the reconciler starts, it will still not detect that the seed was deleted (it reads from an outdated cache).
A shoot resync doesn't help as well, because the reconciler still reads from a stale seed cache.

The only thing that helps in that case is actually a the eventual full relist of the seed informer.
This will emit delete events for objects in the cache that are not included in the list of the relist operation, see [delta_fifo.go](https://github.com/kubernetes/client-go/blob/a4e27104a2f8f5d3965f1520f74a190eba1ad221/tools/cache/delta_fifo.go#L598).
Hence, the only way to miss a delete event is by the controller going down. But after a restart, we will enqueue all shoots and accordingly remove seed conditions if the seed object is gone.

I admit, it's pretty hard to wrap my head around this. 
What about simply respecting resync events for both informers to be on the safe side? Shouldn't hurt, right?

Reply:
Sounds good to me 👍

Reply:
> Hence, the only way to miss a delete event is by the controller going down. But after a restart, we will enqueue all shoots and accordingly remove seed conditions if the seed object is gone. 

Doesn't that mean we are already safe and sound with the current solution?
I am hesitant to add more code for a potential problem happening on seed deletion. Especially as Seed deletion is such a corner case. But if you insist I will add it. WDYT?

Reply:
Fine for me, then leave it as it.